pkg/slsa/attest.go

```diff
@@ -0,0 +1,258 @@
+package slsa
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsacommon "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/common"
+	slsav1 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	buildv1alpha2 "github.com/pivotal/kpack/pkg/apis/build/v1alpha2"
+	"github.com/pivotal/kpack/pkg/cnb"
+	"github.com/pivotal/kpack/pkg/config"
+)
+
+type BuilderID string
+
+const (
+	SignedBuildID   BuilderID = "https://kpack.io/slsa/signed-build"
+	UnsignedBuildID BuilderID = "https://kpack.io/slsa/unsigned-build"
+	MediaTypeJSON             = "application/json"
+)
+
+type LifecycleProvider interface {
+	Metadata() (cnb.LifecycleMetadata, error)
+}
+
+type ImageReader interface {
+	Read(keychain authn.Keychain, repoName string) (string, string, map[string]string, error)
+}
+
+type Attester struct {
+	Version string
+
+	ImageReader       ImageReader
+	LifecycleProvider LifecycleProvider
+
+	Images   config.Images
+	Features config.FeatureFlags
+	Config   config.Config
+}
+
+func (a *Attester) AttestBuild(build *buildv1alpha2.Build, buildMetadata *cnb.BuildMetadata, pod *corev1.Pod, builderAndAppKeychain authn.Keychain, builderId BuilderID, depFns ...BuilderDependencyFn) (intoto.Statement, error) {
+	builderRepo, builderSha, builderLabels, err := a.ImageReader.Read(builderAndAppKeychain, build.Spec.Builder.Image)
+	if err != nil {
+		return intoto.Statement{}, fmt.Errorf("reading builder image: %v", err)
+	}
+
+	appRepo, appSha, appLabels, err := a.ImageReader.Read(builderAndAppKeychain, buildMetadata.LatestImage)
+	if err != nil {
+		return intoto.Statement{}, fmt.Errorf("reading app image: %v", err)
+	}
+
+	source, sourceDigest, err := extractSourceFromLabel(appLabels)
+	if err != nil {
+		return intoto.Statement{}, fmt.Errorf("extracting source from label: %v", err)
+	}
+
+	start, stop, err := getStartStopTime(pod)
```

Comment:
I don't expect this to ever fail, but it _appears_ that the `StartedOn`/`FinishedOn` is optional in https://pkg.go.dev/github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1#BuildMetadata.  Should this failure stop the attestation process?

Reply:
Good catch, I've made `getStartStopTime` do a best effort in parsing and have no failure states